File: app/src/main/java/com/qrdrop/phototransfer/ui/ReceiverScreen.kt
package com.qrdrop.phototransfer.ui

import android.Manifest
import android.content.contentValuesOf
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.widget.Toast
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CameraAlt
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.Save
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.content.ContextCompat
import com.qrdrop.phototransfer.ui.theme.DarkBackground
import com.qrdrop.phototransfer.ui.theme.DarkSurface
import com.qrdrop.phototransfer.ui.theme.DarkSurfaceVariant
import com.qrdrop.phototransfer.ui.theme.MutedText
import com.qrdrop.phototransfer.ui.theme.OnDarkText
import com.qrdrop.phototransfer.ui.theme.PrimaryCyan
import com.qrdrop.phototransfer.ui.theme.SecondaryGreen
import com.qrdrop.phototransfer.viewmodel.ReceiverViewModel

@Composable
fun ReceiverScreen(
    viewModel: ReceiverViewModel,
    modifier: Modifier = Modifier
) {
    val context = LocalContext.current
    val uiState by viewModel.uiState.collectAsState()

    var hasCameraPermission by remember {
        mutableStateOf(
            ContextCompat.checkSelfPermission(
                context,
                Manifest.permission.CAMERA
            ) == PackageManager.PERMISSION_GRANTED
        )
    }

    val permissionLauncher = rememberLauncherForActivityResult(
        contract = ActivityResultContracts.RequestPermission()
    ) { isGranted ->
        hasCameraPermission = isGranted
    }

    LaunchedEffect(uiState.saveResult) {
        uiState.saveResult?.let { resultMsg ->
            Toast.makeText(context, resultMsg, Toast.LENGTH_LONG).show()
        }
    }

    Surface(
        modifier = modifier.fillMaxSize(),
        color = DarkBackground
    ) {
        if (!hasCameraPermission) {
            CameraPermissionPrompt(
                onRequestPermission = {
                    permissionLauncher.launch(Manifest.permission.CAMERA)
                }
            )
        } else {
            Box(modifier = Modifier.fillMaxSize()) {
                Column(
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(16.dp),
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    // Header Bar
                    HeaderSection(uiState.isCompleted)

                    Spacer(modifier = Modifier.height(12.dp))

                    // Live Photo Reconstruction Canvas
                    Box(
                        modifier = Modifier
                            .weight(1f)
                            .fillMaxWidth()
                            .clip(RoundedCornerShape(16.dp))
                            .background(DarkSurface)
                            .border(1.dp, DarkSurfaceVariant, RoundedCornerShape(16.dp)),
                        contentAlignment = Alignment.Center
                    ) {
                        val currentBitmap = uiState.bitmap
                        if (currentBitmap != null) {
                            // Draw the actual bitmap being live reconstructed
                            // Keying on uiState.version guarantees instant Compose updates as frames arrive
                            val imageBitmap = remember(uiState.version) { currentBitmap.asImageBitmap() }
                            Image(
                                bitmap = imageBitmap,
                                contentDescription = "Live Photo Canvas",
                                modifier = Modifier
                                    .fillMaxSize()
                                    .padding(8.dp),
                                contentScale = ContentScale.Fit
                            )
                        } else {
                            // Blank/Placeholder state before first valid QR packet
                            InitialPlaceholder()
                        }
                    }

                    Spacer(modifier = Modifier.height(16.dp))

                    // Progress & Telemetry Panel
                    ProgressTelemetryPanel(
                        receivedFrames = uiState.receivedFramesCount,
                        totalFrames = uiState.totalFramesCount,
                        percentage = uiState.progressPercentage,
                        width = uiState.width,
                        height = uiState.height,
                        statusMessage = uiState.statusMessage,
                        isCompleted = uiState.isCompleted,
                        isSaving = uiState.isSaving,
                        onSaveClicked = { viewModel.saveToGallery(context) },
                        onResetClicked = { viewModel.reset() }
                    )
                }

                // Secondary Small Floating Camera Preview (PIP mode)
                if (uiState.isScanning) {
                    Box(
                        modifier = Modifier
                            .align(Alignment.BottomEnd)
                            .padding(end = 24.dp, bottom = 120.dp)
                            .size(width = 110.dp, height = 150.dp)
                            .clip(RoundedCornerShape(12.dp))
                            .background(Color.Black)
                            .border(2.dp, PrimaryCyan, RoundedCornerShape(12.dp))
                    ) {
                        CameraPreviewView(
                            onBarcodeDetected = { rawQr ->
                                viewModel.processQrText(rawQr)
                            },
                            modifier = Modifier.fillMaxSize()
                        )

                        // Active Scanner Indicator Badge
                        Row(
                            modifier = Modifier
                                .align(Alignment.TopStart)
                                .padding(6.dp)
                                .clip(RoundedCornerShape(4.dp))
                                .background(Color.Black.copy(alpha = 0.7f))
                                .padding(horizontal = 4.dp, vertical = 2.dp),
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Box(
                                modifier = Modifier
                                    .size(6.dp)
                                    .clip(CircleShape)
                                    .background(SecondaryGreen)
                            )
                            Spacer(modifier = Modifier.width(4.dp))
                            Text(
                                text = "SCANNING",
                                fontSize = 8.sp,
                                fontWeight = FontWeight.Bold,
                                color = Color.White
                            )
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun HeaderSection(isCompleted: Boolean) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Column {
            Text(
                text = "QR Photo Transfer",
                fontSize = 22.sp,
                fontWeight = FontWeight.Bold,
                color = OnDarkText
            )
            Text(
                text = "Offline Live Optical Receiver",
                fontSize = 12.sp,
                color = MutedText
            )
        }

        Box(
            modifier = Modifier
                .clip(RoundedCornerShape(20.dp))
                .background(if (isCompleted) SecondaryGreen.copy(alpha = 0.2f) else PrimaryCyan.copy(alpha = 0.2f))
                .border(
                    1.dp,
                    if (isCompleted) SecondaryGreen else PrimaryCyan,
                    RoundedCornerShape(20.dp)
                )
                .padding(horizontal = 12.dp, vertical = 6.dp)
        ) {
            Text(
                text = if (isCompleted) "COMPLETE" else "LIVE SCAN",
                fontSize = 11.sp,
                fontWeight = FontWeight.Bold,
                color = if (isCompleted) SecondaryGreen else PrimaryCyan
            )
        }
    }
}

@Composable
private fun InitialPlaceholder() {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
        modifier = Modifier.padding(24.dp)
    ) {
        Icon(
            imageVector = Icons.Default.CameraAlt,
            contentDescription = null,
            modifier = Modifier.size(56.dp),
            tint = PrimaryCyan.copy(alpha = 0.6f)
        )
        Spacer(modifier = Modifier.height(16.dp))
        Text(
            text = "Waiting for QR Photo Stream...",
            fontSize = 16.sp,
            fontWeight = FontWeight.SemiBold,
            color = OnDarkText,
            textAlign = TextAlign.Center
        )
        Spacer(modifier = Modifier.height(8.dp))
        Text(
            text = "Point camera at the sender screen. Pixels will reconstruct live in real-time.",
            fontSize = 12.sp,
            color = MutedText,
            textAlign = TextAlign.Center
        )
    }
}

@Composable
private fun ProgressTelemetryPanel(
    receivedFrames: Int,
    totalFrames: Int,
    percentage: Float,
    width: Int,
    height: Int,
    statusMessage: String,
    isCompleted: Boolean,
    isSaving: Boolean,
    onSaveClicked: () -> Unit,
    onResetClicked: () -> Unit
) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        shape = RoundedCornerShape(16.dp),
        colors = CardDefaults.cardColors(containerColor = DarkSurface)
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp)
        ) {
            if (isCompleted) {
                // Completion Banner & Actions
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Icon(
                        imageVector = Icons.Default.CheckCircle,
                        contentDescription = null,
                        tint = SecondaryGreen,
                        modifier = Modifier.size(28.dp)
                    )
                    Spacer(modifier = Modifier.width(10.dp))
                    Column(modifier = Modifier.weight(1f)) {
                        Text(
                            text = "Transfer Complete",
                            fontSize = 16.sp,
                            fontWeight = FontWeight.Bold,
                            color = SecondaryGreen
                        )
                        Text(
                            text = "$receivedFrames / $totalFrames frames ($width x $height px)",
                            fontSize = 12.sp,
                            color = MutedText
                        )
                    }
                }

                Spacer(modifier = Modifier.height(12.dp))

                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.spacedBy(12.dp)
                ) {
                    Button(
                        onClick = onSaveClicked,
                        enabled = !isSaving,
                        modifier = Modifier.weight(1f),
                        shape = RoundedCornerShape(10.dp),
                        colors = ButtonDefaults.buttonColors(containerColor = SecondaryGreen, contentColor = Color.Black)
                    ) {
                        if (isSaving) {
                            CircularProgressIndicator(
                                modifier = Modifier.size(18.dp),
                                color = Color.Black,
                                strokeWidth = 2.dp
                            )
                        } else {
                            Icon(imageVector = Icons.Default.Save, contentDescription = null, modifier = Modifier.size(18.dp))
                            Spacer(modifier = Modifier.width(8.dp))
                            Text(text = "Save to Gallery", fontWeight = FontWeight.Bold)
                        }
                    }

                    OutlinedButton(
                        onClick = onResetClicked,
                        shape = RoundedCornerShape(10.dp),
                        colors = ButtonDefaults.outlinedButtonColors(contentColor = OnDarkText)
                    ) {
                        Icon(imageVector = Icons.Default.Refresh, contentDescription = null, modifier = Modifier.size(18.dp))
                    }
                }
            } else {
                // Live Transfer Progress
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        text = if (totalFrames > 0) "$receivedFrames / $totalFrames frames" else "0 / 0 frames",
                        fontSize = 14.sp,
                        fontWeight = FontWeight.Medium,
                        color = OnDarkText
                    )
                    Text(
                        text = String.format("%.1f%%", percentage),
                        fontSize = 16.sp,
                        fontWeight = FontWeight.Bold,
                        fontFamily = FontFamily.Monospace,
                        color = PrimaryCyan
                    )
                }

                Spacer(modifier = Modifier.height(8.dp))

                LinearProgressIndicator(
                    progress = { if (totalFrames > 0) receivedFrames.toFloat() / totalFrames else 0f },
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(8.dp)
                        .clip(RoundedCornerShape(4.dp)),
                    color = PrimaryCyan,
                    trackColor = DarkSurfaceVariant
                )

                Spacer(modifier = Modifier.height(10.dp))

                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    Text(
                        text = "Status: $statusMessage",
                        fontSize = 12.sp,
                        color = MutedText
                    )
                    if (width > 0 && height > 0) {
                        Text(
                            text = "${width}x${height}px",
                            fontSize = 12.sp,
                            color = MutedText,
                            fontFamily = FontFamily.Monospace
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun CameraPermissionPrompt(
    onRequestPermission: () -> Unit
) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ) {
        Icon(
            imageVector = Icons.Default.CameraAlt,
            contentDescription = null,
            modifier = Modifier.size(72.dp),
            tint = PrimaryCyan
        )
        Spacer(modifier = Modifier.height(24.dp))
        Text(
            text = "Camera Permission Required",
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold,
            color = OnDarkText,
            textAlign = TextAlign.Center
        )
        Spacer(modifier = Modifier.height(12.dp))
        Text(
            text = "QR Photo Transfer needs camera access to continuously scan QR photo streams and reconstruct your photos offline.",
            fontSize = 14.sp,
            color = MutedText,
            textAlign = TextAlign.Center
        )
        Spacer(modifier = Modifier.height(28.dp))
        Button(
            onClick = onRequestPermission,
            shape = RoundedCornerShape(12.dp),
            colors = ButtonDefaults.buttonColors(containerColor = PrimaryCyan, contentColor = Color.Black)
        ) {
            Text(text = "Grant Camera Permission", fontWeight = FontWeight.Bold)
        }
    }
}

File: lib/ui/receiver_screen.dart
import 'package:flutter/material.dart';
import 'package:mobile_scanner/mobile_scanner.dart';
import 'package:permission_handler/permission_handler.dart';
import '../engine/image_reconstruction_engine.dart';
import '../parser/packet_parser.dart';
import '../saver/gallery_saver.dart';

class ReceiverScreen extends StatefulWidget {
  const ReceiverScreen({super.key});

  @override
  State<ReceiverScreen> createState() => _ReceiverScreenState();
}

class _ReceiverScreenState extends State<ReceiverScreen> {
  final ImageReconstructionEngine _engine = ImageReconstructionEngine();
  final MobileScannerController _scannerController = MobileScannerController(
    detectionSpeed: DetectionSpeed.unrestricted,
    returnImage: false,
  );

  bool _isScanning = true;
  bool _isSaving = false;
  bool _hasCameraPermission = false;
  String _statusMessage = 'Scanning for QR stream...';

  @override
  void initState() {
    super.initState();
    _checkPermission();
  }

  Future<void> _checkPermission() async {
    final status = await Permission.camera.status;
    if (status.isGranted) {
      setState(() {
        _hasCameraPermission = true;
      });
    } else {
      final requestStatus = await Permission.camera.request();
      setState(() {
        _hasCameraPermission = requestStatus.isGranted;
      });
    }
  }

  void _onBarcodeDetected(BarcodeCapture capture) async {
    if (!_isScanning || _engine.isComplete) return;

    for (final barcode in capture.barcodes) {
      final rawValue = barcode.rawValue;
      if (rawValue == null || rawValue.isEmpty) continue;

      final packet = PacketParser.parse(rawValue);
      if (packet == null) continue;

      final isNewChunk = await _engine.processPacket(packet);
      if (isNewChunk && mounted) {
        setState(() {
          if (_engine.isComplete) {
            _isScanning = false;
            _statusMessage = _engine.hasDecodeError
                ? 'JPEG decode error'
                : 'Transfer Complete!';
          } else {
            _statusMessage =
                'Receiving JPEG chunk ${packet.frameIndex + 1} / ${_engine.totalFrames}';
          }
        });
      }
    }
  }

  Future<void> _saveToGallery() async {
    final jpegBytes = _engine.completeJpegBytes;
    if (jpegBytes == null || _isSaving) return;

    setState(() {
      _isSaving = true;
    });

    final result = await GallerySaver.saveToGallery(
      jpegBytes: jpegBytes,
    );

    if (mounted) {
      setState(() {
        _isSaving = false;
      });

      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text(result ?? 'Saved to Gallery')),
      );
    }
  }

  void _reset() {
    _engine.reset();
    setState(() {
      _isScanning = true;
      _isSaving = false;
      _statusMessage = 'Scanning for QR stream...';
    });
  }

  @override
  void dispose() {
    _scannerController.dispose();
    _engine.reset();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: const Color(0xFF101216),
      body: SafeArea(
        child: Stack(
          children: [
            Column(
              children: [
                // Header Bar
                _buildHeader(),

                const SizedBox(height: 12),

                // Main Photo Reconstruction Display
                Expanded(
                  child: Container(
                    margin: const EdgeInsets.symmetric(horizontal: 16),
                    decoration: BoxDecoration(
                      color: const Color(0xFF1A1D24),
                      borderRadius: BorderRadius.circular(16),
                      border: Border.all(color: const Color(0xFF252932)),
                    ),
                    clipBehavior: Clip.antiAlias,
                    child: Center(
                      child: _engine.uiImage != null
                          ? RawImage(
                              image: _engine.uiImage,
                              fit: BoxFit.contain,
                            )
                          : _engine.hasDecodeError
                              ? _buildErrorDisplay()
                              : _buildInitialPlaceholder(),
                    ),
                  ),
                ),

                const SizedBox(height: 16),

                // Telemetry & Progress Panel
                _buildProgressPanel(),
              ],
            ),

            // Secondary Small Floating Camera Preview (PIP mode)
            if (_isScanning && _hasCameraPermission)
              Positioned(
                right: 24,
                bottom: 120,
                child: Container(
                  width: 110,
                  height: 150,
                  decoration: BoxDecoration(
                    color: Colors.black,
                    borderRadius: BorderRadius.circular(12),
                    border: Border.all(color: const Color(0xFF00E5FF), width: 2),
                  ),
                  clipBehavior: Clip.antiAlias,
                  child: Stack(
                    children: [
                      MobileScanner(
                        controller: _scannerController,
                        onDetect: _onBarcodeDetected,
                        errorBuilder: (context, error, child) {
                          return const Center(
                            child: Icon(Icons.camera_alt_outlined, color: Colors.white54, size: 28),
                          );
                        },
                      ),
                      Positioned(
                        top: 6,
                        left: 6,
                        child: Container(
                          padding: const EdgeInsets.symmetric(horizontal: 4, vertical: 2),
                          decoration: BoxDecoration(
                            color: Colors.black.withAlpha(180),
                            borderRadius: BorderRadius.circular(4),
                          ),
                          child: const Row(
                            mainAxisSize: MainAxisSize.min,
                            children: [
                              CircleAvatar(
                                radius: 3,
                                backgroundColor: Color(0xFF00E676),
                              ),
                              SizedBox(width: 4),
                              Text(
                                'SCANNING',
                                style: TextStyle(
                                  fontSize: 8,
                                  fontWeight: FontWeight.bold,
                                  color: Colors.white,
                                ),
                              ),
                            ],
                          ),
                        ),
                      ),
                    ],
                  ),
                ),
              ),

            if (!_hasCameraPermission)
              Positioned.fill(
                child: Container(
                  color: const Color(0xFF101216),
                  padding: const EdgeInsets.all(24),
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      const Icon(Icons.camera_alt, size: 64, color: Color(0xFF00E5FF)),
                      const SizedBox(height: 16),
                      const Text(
                        'Camera Permission Required',
                        style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold, color: Colors.white),
                      ),
                      const SizedBox(height: 8),
                      const Text(
                        'Camera access is needed to scan compressed-JPEG QR streams.',
                        textAlign: TextAlign.center,
                        style: TextStyle(fontSize: 13, color: Color(0xFF90A4AE)),
                      ),
                      const SizedBox(height: 24),
                      ElevatedButton(
                        onPressed: _checkPermission,
                        style: ElevatedButton.styleFrom(
                          backgroundColor: const Color(0xFF00E5FF),
                          foregroundColor: Colors.black,
                        ),
                        child: const Text('Grant Camera Permission', style: TextStyle(fontWeight: FontWeight.bold)),
                      ),
                    ],
                  ),
                ),
              ),
          ],
        ),
      ),
    );
  }

  Widget _buildHeader() {
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: [
          const Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text(
                'QR Photo Transfer',
                style: TextStyle(
                  fontSize: 22,
                  fontWeight: FontWeight.bold,
                  color: Color(0xFFECEFF1),
                ),
              ),
              Text(
                'Compressed-JPEG QR Receiver',
                style: TextStyle(
                  fontSize: 12,
                  color: Color(0xFF90A4AE),
                ),
              ),
            ],
          ),
          Container(
            padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 6),
            decoration: BoxDecoration(
              color: _engine.isComplete
                  ? const Color(0xFF00E676).withAlpha(50)
                  : const Color(0xFF00E5FF).withAlpha(50),
              borderRadius: BorderRadius.circular(20),
              border: Border.all(
                color: _engine.isComplete ? const Color(0xFF00E676) : const Color(0xFF00E5FF),
              ),
            ),
            child: Text(
              _engine.isComplete ? 'COMPLETE' : 'LIVE SCAN',
              style: TextStyle(
                fontSize: 11,
                fontWeight: FontWeight.bold,
                color: _engine.isComplete ? const Color(0xFF00E676) : const Color(0xFF00E5FF),
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildInitialPlaceholder() {
    return Padding(
      padding: const EdgeInsets.all(24.0),
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          const Icon(
            Icons.camera_alt,
            size: 56,
            color: Color(0x9900E5FF),
          ),
          const SizedBox(height: 16),
          const Text(
            'Waiting for Compressed-JPEG QR Stream...',
            style: TextStyle(
              fontSize: 16,
              fontWeight: FontWeight.w600,
              color: Color(0xFFECEFF1),
            ),
            textAlign: TextAlign.center,
          ),
          const SizedBox(height: 8),
          Text(
            _engine.totalFrames > 0
                ? 'Receiving JPEG chunks (${_engine.receivedFrameCount} / ${_engine.totalFrames})...'
                : 'Point camera at sender screen. Full photo will decode when all chunks arrive.',
            style: const TextStyle(
              fontSize: 12,
              color: Color(0xFF90A4AE),
            ),
            textAlign: TextAlign.center,
          ),
        ],
      ),
    );
  }

  Widget _buildErrorDisplay() {
    return const Padding(
      padding: EdgeInsets.all(24.0),
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Icon(
            Icons.error_outline,
            size: 56,
            color: Colors.redAccent,
          ),
          SizedBox(height: 16),
          Text(
            'JPEG Decoding Error',
            style: TextStyle(
              fontSize: 16,
              fontWeight: FontWeight.w600,
              color: Colors.redAccent,
            ),
            textAlign: TextAlign.center,
          ),
          SizedBox(height: 8),
          Text(
            'The reconstructed JPEG data is corrupted or incomplete. Please reset and scan again.',
            style: TextStyle(
              fontSize: 12,
              color: Color(0xFF90A4AE),
            ),
            textAlign: TextAlign.center,
          ),
        ],
      ),
    );
  }

  Widget _buildProgressPanel() {
    final isComplete = _engine.isComplete;
    return Container(
      margin: const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
      padding: const EdgeInsets.all(16),
      decoration: BoxDecoration(
        color: const Color(0xFF1A1D24),
        borderRadius: BorderRadius.circular(16),
      ),
      child: Column(
        children: [
          if (isComplete && !_engine.hasDecodeError) ...[
            const Row(
              children: [
                Icon(Icons.check_circle, color: Color(0xFF00E676), size: 28),
                SizedBox(width: 10),
                Expanded(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        'Transfer Complete',
                        style: TextStyle(
                          fontSize: 16,
                          fontWeight: FontWeight.bold,
                          color: Color(0xFF00E676),
                        ),
                      ),
                      Text(
                        'JPEG photo decoded successfully',
                        style: TextStyle(fontSize: 12, color: Color(0xFF90A4AE)),
                      ),
                    ],
                  ),
                ),
              ],
            ),
            const SizedBox(height: 12),
            Row(
              children: [
                Expanded(
                  child: ElevatedButton.icon(
                    onPressed: _isSaving ? null : _saveToGallery,
                    icon: _isSaving
                        ? const SizedBox(
                            width: 18,
                            height: 18,
                            child: CircularProgressIndicator(strokeWidth: 2, color: Colors.black),
                          )
                        : const Icon(Icons.save, size: 18),
                    label: const Text('Save to Gallery', style: TextStyle(fontWeight: FontWeight.bold)),
                    style: ElevatedButton.styleFrom(
                      backgroundColor: const Color(0xFF00E676),
                      foregroundColor: Colors.black,
                      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(10)),
                    ),
                  ),
                ),
                const SizedBox(width: 12),
                OutlinedButton(
                  onPressed: _reset,
                  style: OutlinedButton.styleFrom(
                    foregroundColor: const Color(0xFFECEFF1),
                    shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(10)),
                  ),
                  child: const Icon(Icons.refresh, size: 18),
                ),
              ],
            ),
          ] else ...[
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Text(
                  _engine.totalFrames > 0
                      ? '${_engine.receivedFrameCount} / ${_engine.totalFrames} chunks'
                      : '0 / 0 chunks',
                  style: const TextStyle(fontSize: 14, color: Color(0xFFECEFF1)),
                ),
                Text(
                  '${_engine.progressPercentage.toStringAsFixed(1)}%',
                  style: const TextStyle(
                    fontSize: 16,
                    fontWeight: FontWeight.bold,
                    fontFamily: 'monospace',
                    color: Color(0xFF00E5FF),
                  ),
                ),
              ],
            ),
            const SizedBox(height: 8),
            LinearProgressIndicator(
              value: _engine.totalFrames > 0 ? _engine.receivedFrameCount / _engine.totalFrames : 0,
              backgroundColor: const Color(0xFF252932),
              color: _engine.hasDecodeError ? Colors.redAccent : const Color(0xFF00E5FF),
              minHeight: 8,
              borderRadius: BorderRadius.circular(4),
            ),
            const SizedBox(height: 10),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Text(
                  'Status: $_statusMessage',
                  style: const TextStyle(fontSize: 12, color: Color(0xFF90A4AE)),
                ),
                if (_engine.width > 0)
                  Text(
                    '${_engine.width}x${_engine.height}px',
                    style: const TextStyle(fontSize: 12, color: Color(0xFF90A4AE), fontFamily: 'monospace'),
                  ),
              ],
            ),
          ],
        ],
      ),
    );
  }
}
